src/ray/common/asio/periodical_runner.cc

```diff
@@ -52,4 +59,30 @@ void PeriodicalRunner::DoRunFnPeriodically(std::function<void()> fn,
   });
 }
 
+void PeriodicalRunner::DoRunFnPeriodicallyInstrumented(
+    std::function<void()> fn, boost::posix_time::milliseconds period,
+    boost::asio::deadline_timer &timer, const std::string name) {
+  fn();
+  timer.expires_from_now(period);
+  // NOTE: We add the timer period to the enqueue time in order only measure the time in
+  // which the handler was elgible to execute on the event loop but was queued by the
+  // event loop.
+  auto stats_handle = io_service_.RecordStart(name, period.total_nanoseconds());
```

Comment:
Could we drop this padding? I think it's weird to include the timer delay, and it will confuse the global states.

Reply:
This is padding the enqueue start time, not the queue time duration, i.e. this is ensuring that the timer delay is _not_ included in the queue time.

Reply:
Oh I see, it's subtracting... hmm how about not using RecordStart for timers then? We probably don't need to know the timer is queued, since at most 1 execution can be queued at once.

Reply:
Hmm even though only one timer handler per timer can be queued at a time (so the current count will always be 0 or 1 for each timer handler), `RecordStart()` is still used for measuring event loop queueing delays, which are globally aggregated across all handlers that are submitted to the event loop, so I'd still like for the deadline timer handlers to be included in that calculation. And not using `RecordStart()` will just add complexity to the implementation of `RecordExecution()`.

Reply:
Wouldn't removing this also allow us to fix the issue you mention above?

> Hmm I just implemented this, and it appears that this won't behave well with deadline timers. If the periodical runner is destructed or the deadline timer is cancelled, the callback will never be invoked and the opaque stats handle will be destructed without RecordExecution having been called.

I think this is exactly the kind of edge case we want to handle--- cancellation that causes us to forget to decrement the active operations counter. It would be great to try to get this fix in otherwise we might leak operations on cancel quite easily.

Reply:
I think that I have a fix for this that prevents these leaks while still fully instrumenting the deadline timers.